src/dataframerow/utils.jl

```diff
@@ -145,50 +146,75 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
     return ngroups, rhashes, gslots, false
 end
 
-function row_group_slots(cols::Tuple{CategoricalVector},
+function row_group_slots(cols::NTuple{N,<:CategoricalVector},
                          hash::Val{false},
                          groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool}
-    col = cols[1]
-    @assert groups === nothing || length(groups) == length(col)
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool} where N
+    # Computing neither hashes nor groups isn't very useful,
+    # and this method needs to allocate a groups vector anyway
+    @assert groups !== nothing && all(col -> length(col) == length(groups), cols)
 
     # If missings are to be skipped, they will all go to group 1,
     # which will be removed by group_rows
-    ngroups = length(levels(col)) + (eltype(col) >: Missing)
+    ngroupstup = map(cols) do c
+        length(levels(c)) + (!skipmissing && eltype(c) >: Missing)
+    end
+    ngroups = prod(ngroupstup) + skipmissing
+
+    # Fall back to hashing if there would be too many empty combinations.
+    # The first check ensures the computation of ngroups did not overflow.
+    # The rationale for the 2 threshold is that while the fallback method is always slower,
+    # it allocates a hash table of size length(groups) instead of the remap vector
+    # of size ngroups (i.e. the number of possible combinations) in this method:
+    # so it makes sense to allocate more memory for better performance,
+    # but it needs to remain reasonable compared with the size of the data frame.
```

Comment:
is `2` a guess or you did the benchmarks (I do not insist to do them - just asking)

Reply:
I used the benchmark in the description, varying the number of groups, and it appears the optimized method is always faster (which is why I noted that in the comment). But of course it's not reasonable to allocate 20 or 100 times more memory just to be a bit faster (anyway this case should be pretty rare).

Reply:
Using a `BitArray` would indeed reduce allocations from `2 * ngroups` to `ngroups + ngroups/8`. But I'm not sure it's a good trade-off since CategoricalArrays are optimized for small number of groups: probably better be as fast as possible, in particular in the most common case where the number of combinations is well below twice the number of rows.

Reply:
OK